src/core/slicer.sol

```diff
@@ -0,0 +1,145 @@
+// Copyright (C) 2019  Centrifuge
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.4.24;
+
+import "ds-note/note.sol";
+
+// Slicer
+//  Calculates payouts and slices based on the current iSupply chi. Manages iSupply.
+contract Slicer is DSNote {
+    // --- Auth ---
+    mapping (address => uint) public wards;
+    function rely(address usr) public auth note { wards[usr] = 1; }
+    function deny(address usr) public auth note { wards[usr] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    struct Fee {
+        uint chi;
+        uint speed;
+        uint48 rho;
+    }
+    
+    Fee public iSupply;
+    
+    constructor() public {
+        wards[msg.sender] = 1;
+        iSupply.chi = ONE;
+        iSupply.speed = ONE;
+    }
+
+    function file(bytes32 what, uint speed_) public note auth {
+        if (what == "isupply") { 
+            drip();
+            iSupply.speed = speed_;    
+        } 
+    }
+
+    function chop(uint wadT) public note auth returns (uint) {
+        if (now > iSupply.rho) {
+            drip();
+        }
+        uint slice = rdiv(wadT, iSupply.chi);
+        return slice;
+    }
+
+    function payout(uint wadS) public note auth returns (uint) {
```

Comment:
I thought we will implement the `percentage payout` formula?

In my opinion the following approach `tokenAmount * chi` is not covering losses.

Reply:
this implementation does not consider losses yet.